Add justification to batches, progress on validation pipeline

nomaxg opened the pull request:
**This PR introduces:**
- justification to Arbitrum batches which will be validated against the HotShot commitment corresponding to the batch
- A series of host-io changes to the wasm validation pipeline that allows the wasm executable to read a hotshot commitment (currently, these default to [0; 32] in many places to ensure backwards compatibility)

**Future work:**
- Tests
- Add logic to fetch the HotShot commitment from a live contract (this is currently mocked)
- Set Espresso flags correctly in the espresso variant of nitro-testnode

**Depends on:**
https://github.com/EspressoSystems/geth-arbitrum/pull/2

**To test (for backwards compatibility):**
```bash
./test-node.bash --dev --init
./test-node.bash script send-l2 --to address_0x673C9b46952080D3bf1B8d689C7BBC590a8755Cb --ethamount 5 --l2url
# spin up validation nodes
docker compose up validation_node
docker compose up validator
```

nomaxg committed:
work on batch validation

nomaxg committed:
add input for last hotshot header

nomaxg committed:
progress with header interface

nomaxg committed:
stub out preimage fetching

nomaxg commented on arbos/block_processor.go:
```diff
@@ -168,6 +170,7 @@ func ProduceBlockAdvanced(
 	txes types.Transactions,
 	delayedMessagesRead uint64,
 	lastBlockHeader *types.Header,
+	lastHotShotHeader *espresso.Header,
```
1. `lastHotShotHeader` is added as an input to this method, which is the bulk of the "State Transition Function". Now, the mock namespace proof can be verified against the provided header (below I compare the provided preimage header to the header in the justification, but I guess we don't necessarily need the header in the justification at all if we do things this way).

jbearer replied:
Re: not putting the header in the justification: we could do that, but it would kind of break the "L1 DA" property, where everything you need to validate the L2 chain is available on L1.

If we were to do this, we could probably remove the justification completely and just have the validator fetch whatever proof it needs from HotShot and provide it as a hint to the WASM verifier.

nomaxg replied:
Yes, we assume 1:1 here, last probably isn't the best word, it's more like the HotShotSafeHead

nomaxg commented on cmd/replay/main.go:
```diff
@@ -46,6 +47,19 @@ func getBlockHeaderByHash(hash common.Hash) *types.Header {
 	return header
 }
 
+func getHotShotHeaderByHash(hash common.Hash) *espresso.Header {
```
2. The replay executable is modified to be able to fetch a hotshot header preimage by hash, so that it can be passed into the STF.

nomaxg commented on staker/stateless_block_validator.go:
```diff
@@ -69,6 +70,7 @@ type TransactionStreamerInterface interface {
 
 type InboxReaderInterface interface {
 	GetSequencerMessageBytes(ctx context.Context, seqNum uint64) ([]byte, error)
+	FetchHotShotCommitment(blockHeight uint64) (espresso.Commitment, error)
```
3. We can add a method like this to fetch the desired hotshot commitment in the validation pipeline, which will end up in the ValidationInput passed to the WASM layer.

nomaxg commented on staker/stateless_block_validator.go:
```diff
@@ -305,6 +308,13 @@ func (v *StatelessBlockValidator) ValidationEntryRecord(ctx context.Context, e *
 				return err
 			}
 		}
+		if usingEspresso {
```
4. Finally, we can add the preimage here by calling the inbox oracle method we added. If the HotShot header isn't available, `e.State` will not be set to ready below. This may be how validation is delayed. Not sure about the specific retry logic yet.

jbearer commented on cmd/replay/main.go:
```diff
@@ -46,6 +47,19 @@ func getBlockHeaderByHash(hash common.Hash) *types.Header {
+		return nil
+	}
+	var header &espresso.Header
+	if err := json.Unmarshal(header, &enc); err != nil {
```
I'm confused by this. If the preimage we fetched is a JSON object, that means `hash` is the keccak hash of a JSON string. But usually we don't have the keccak hash of the header serialized as JSON. Usually we have the Hotshot block commitment, which is an incremental keccak hash of all the fields in the header

nomaxg replied:
Yes still trying to wrap my head around this preimage business, as @sveitser  suggested, might be better entirely to treat hotshot commitments like delayed inbox messages

nomaxg committed:
dont use preimage

sveitser commented on cmd/replay/main.go:
```diff
@@ -29,15 +29,13 @@ import (
 	"github.com/offchainlabs/nitro/wavmio"
 )
 
 func getBlockHeaderByHash(hash common.Hash) *types.Header {
```
I think this can be reverted to the original code now?

sveitser commented on cmd/replay/main.go:
```diff
@@ -152,6 +160,8 @@ func main() {
 
 	lastBlockHash := wavmio.GetLastBlockHash()
+	lastHotShotHeaderHash := wavmio.GetLastHotShotHeaderHash()
+	lastHotShotHeader := getHotShotHeaderByHash(lastHotShotHeaderHash)
```
What do we gain from going via the last hash? Could we have a function `getHotShotHeader()` instead?

sveitser replied:
Hmm. just noticed the types in `wavmio.stub` are all primitive, thinking probably for a reason.

nomaxg replied:
Yeah there are some wasm compilation quirks here I’m still figuring out

nomaxg committed:
compiles, test node still doesn't work

nomaxg committed:
move espresso logic to arbos, handle nil justifications

nomaxg committed:
more progress

nomaxg committed:
progress on custom hotshot header host call

nomaxg committed:
backwards compatability with nitro testnode

nomaxg committed:
more progress on header, now failing to deserialize

nomaxg committed:
switch from header to commitment for good

nomaxg committed:
cleanup, add todos

nomaxg committed:
Merge branch 'integration' into batch-validation

nomaxg committed:
sp

nomaxg committed:
remove hotshot idx

sveitser committed:
Update geth with espresso flag

sveitser reviewed (approved):
LGTM

sveitser commented:
I'm getting this failure locally for `make test-go`

```
=== Failed
=== FAIL: broadcaster ExampleBroadcastMessage_broadcastfeedmessage (0.00s)
=== RUN   ExampleBroadcastMessage_broadcastfeedmessage
got:
{"version":1,"messages":[{"sequenceNumber":12345,"message":{"message":{"header":{"kind":0,"sender":"0x0000000000000000000000000000000000000000","blockNumber":0,"timestamp":0,"requestId":"0x0000000000000000000000000000000000000000000000000000000000000000","baseFeeL1":0,"justification":null},"l2Msg":"3q2+7w=="},"delayedMessagesRead":3333},"signature":null}]}
want:
{"version":1,"messages":[{"sequenceNumber":12345,"message":{"message":{"header":{"kind":0,"sender":"0x0000000000000000000000000000000000000000","blockNumber":0,"timestamp":0,"requestId":"0x0000000000000000000000000000000000000000000000000000000000000000","baseFeeL1":0},"l2Msg":"3q2+7w=="},"delayedMessagesRead":3333},"signature":null}]}

DONE 463 tests, 1 skipped, 1 failure in 117.090s
make: *** [Makefile:327: .make/test-go] Error 1
```

sveitser committed:
Omit empty justification in serializationof L1 msg

This makes the `ExampleBroadcastMessage_broadcastfeedmessage` test pass
without having to change the expected output.

ImJeremyHe commented on arbitrator/jit/src/wavmio.rs:
```diff
@@ -101,6 +109,38 @@ pub fn read_delayed_inbox_message(mut env: WasmEnvMut, sp: u32) -> MaybeEscape {
+    let message = comm_map.get(&msg_num).unwrap_or(&[0; 32]);
+
+    if out_ptr + 32 > sp.memory_size() {
+        error!("memory bounds exceeded in {name}");
```
macro `error!` is used only for once, I think it is better to expand it here and remove the macro.

ImJeremyHe commented on arbos/block_processor.go:
```diff
@@ -191,6 +194,29 @@ func ProduceBlockAdvanced(
 
+	// Espresso-specific validation
+	// TODO test: https://github.com/EspressoSystems/espresso-sequencer/issues/772
+	if chainConfig.Espresso {
```
This seems to conflict with what I am currently doing. I'm inclined to think that conducting Espresso validation might not be suitable in this stage. Perhaps we need to discuss this.
https://github.com/EspressoSystems/nitro-espresso-integration/pull/14

nomaxg replied:
discussed offline, we will move this logic to ProduceBlock where we have access to all transactions

ImJeremyHe commented on arbitrator/jit/src/wavmio.rs:
```diff
@@ -101,6 +109,38 @@ pub fn read_delayed_inbox_message(mut env: WasmEnvMut, sp: u32) -> MaybeEscape {
 }
 
+// Reads a hotshot commitment
+fn read_hotshot_commitment_impl(
```
Could you please help me understand this a bit better?

nomaxg replied:
Yes, essentially what we are doing is validating that espresso batches (which currently map 1-to-1 to arbitrum batches) correspond to a valid hotshot commitment. This function implements a host-io function that reads a hotshot commitment, so that the wasm layer can pass a commitment`ProduceBlock

sveitser committed:
Make go linter happy

nomaxg committed:
expand unecessary macro
